src/rust/engine/src/context.rs

```diff
@@ -120,6 +104,12 @@ impl Context {
   }
 }
 
+impl GetFileDigest<Failure> for Context {
```

Comment:
Remind me again why we don't just put this on `VFS`?

Reply:
I couldn't answer this question, so I tried to move it, and I learnt the answer! Which definitely points to something to improve.

There's a currently-very-implicit side-effect of calling `digest`, which is that after you call it, the `Store` will be able to answer `load_file_bytes` calls for it. This is an important property.

So `VFS` is the wrong place for the `digest` function, because `PosixFS` shouldn't have a dependency on `Store`.

`Store` would be a reasonable place for it, except that `Context` wants to be able to cache it rather than pass through to the underlying `Store`.

So I think we should probably have a standalone trait which is maybe called `StoresFileByDigest`, which `Store` implements (because it does, and is useful in tests), and which `Context` also implements (for the caching layer).

WDYT?

Reply:
... Except `Store` can't implement it on its own, because it needs a VFS to actually load files.

Reply:
Did the rename, added a comment.